### ISSUE TITLE
How to run this project

### ISSUE THREAD

**Issue by JennyWJN** [YoshikiHigo/TinyPDG#4]
Hi, Higo.
I'm trying to run your project on my machine and I'm new to gradle project. I'm not sure where're the input and output. When I try to run the project, I got this.
![image](https://user-images.githubusercontent.com/37525646/107140632-50e59e80-695e-11eb-9527-67af9caecdb4.png)
Can you give me some advice?
Thank you very much!

**Comment by JennyWJN** [comment]
Hi, Higo.
Now I has passed arguments to scorpio, but it seems I gave some wrong value and the output is empty.
This is my Program Arguments.
![image](https://user-images.githubusercontent.com/37525646/107313065-c3748c80-6acc-11eb-8426-68f5b14f3ae6.png)
What should I do?
Thank you!

**Comment by YoshikiHigo** [comment]
Thank you for using TinyPDG.
Would you check if Test001.java includes at least a non-empty method?

**Comment by JennyWJN** [comment]
Yes, because I'm not sure how to run the test files in your project, I copied the method of test001 to this file.
![image](https://user-images.githubusercontent.com/37525646/107333152-085dea80-6af0-11eb-835c-d263914029fb.png)
By the way may I ask how to run those testcases in the test folder?
Thank you!

**Comment by JennyWJN** [comment]
The output is as below.
![image](https://user-images.githubusercontent.com/37525646/107333866-fb8dc680-6af0-11eb-8ebb-f02bf0c19aef.png)

**Comment by YoshikiHigo** [comment]
I see.
The test files are for checking if TinyPDG generates program dependence graphs and not for checking clone detection.
Scorpio is the main class for detecting PDG-based clones.
So, please generate a Java file including two similar methods and then give the file to Scorpio.

**Comment by YoshikiHigo** [comment]
Just now, I've recognized that Scorpio's jar file generated by Gradle does not work.
So, please use Scorpio in your IDE environment.
At this moment, I have no time to modify the gradle file of Scorpio because I need to bring back many things of Scorpio.

**Comment by JennyWJN** [comment]
Hi,
I try this file and get these output.
![image](https://user-images.githubusercontent.com/37525646/107336719-67bdf980-6af4-11eb-8bdb-a4896d202571.png)
![image](https://user-images.githubusercontent.com/37525646/107336743-70aecb00-6af4-11eb-8545-eedd2b772bf6.png)
It seems correct. Thank you very much!
And may I ask if I want to generate PDG as the files in your test folder, is there a way to directly run with the test file? Thank you!

**Comment by JennyWJN** [comment]
Hi, Higo.
I noticed that there's a writer file with main function and solve it. Thank you very much for the help and this tool!

**Comment by QuanixnYang** [comment]
> Hi, Higo.
> I noticed that there's a writer file with main function and solve it. Thank you very much for the help and this tool!



I still don't know the details of using tinyPDG. Can you provide some demonstration, thank you very much.

**Comment by JennyWJN** [comment]
> > Hi, Higo.
> > I noticed that there's a writer file with main function and solve it. Thank you very much for the help and this tool!
> 
> I still don't know the details of using tinyPDG. Can you provide some demonstration, thank you very much.

Hi,
If you want to use the clone detect function, you can run Scorpio file with these program arguments —— “-d target directory -o output file path -s 5(size) -t 1(thread number) -C on(use control dependency) -D on(use date dependency) -E on(use execution dependency) -M on(merging consecutive similar nodes)”. Then you can understand more by observing the output file.

**Comment by QuanixnYang** [comment]
> > > Hi, Higo.
> > > I noticed that there's a writer file with main function and solve it. Thank you very much for the help and this tool!
> > 
> > 
> > I still don't know the details of using tinyPDG. Can you provide some demonstration, thank you very much.
> 
> Hi,
> If you want to use the clone detect function, you can run Scorpio file with these program arguments —— “-d target directory -o output file path -s 5(size) -t 1(thread number) -C on(use control dependency) -D on(use date dependency) -E on(use execution dependency) -M on(merging consecutive similar nodes)”. Then you can understand more by observing the output file.

Thank you very much for your detailed explanation！

**Comment by QuanixnYang** [comment]
Hi, Higo.
I want to get CFG and PDG of Java code. I have tried the Soot tool before, but the effect is not rational. In your project, CFG and PDG have been implemented, and the effect seems to be very good. I just want to analyze CFG and PDG（dot file and png file）, not detect code clone. In addition, I want to output only one CFG or PDG for each code file. Can I merge them by modifying the code in some way? May I get some advice from you? I'm not familiar with Java. Thank you very very much.

**Comment by QuanixnYang** [comment]
> Hi, Higo.
> I want to get CFG and PDG of Java code. I have tried the Soot tool before, but the effect is not rational. In your project, CFG and PDG have been implemented, and the effect seems to be very good. I just want to analyze CFG and PDG（dot file and png file）, not detect code clone. In addition, I want to output only one CFG or PDG for each code file. Can I merge them by modifying the code in some way? May I get some advice from you? I'm not familiar with Java. Thank you very very much.

Ohh, just now,  I find a main() in Writer.java, I successfully got the CFG and PDG dot files. That is Awesome !!
But how can I merge the subgraph clusters? I just want one graph.
![sendpix1](https://user-images.githubusercontent.com/44419438/115245556-a0fb9100-a157-11eb-8e20-7379431df94b.jpg)

**Comment by DanielWang2019** [comment]
If I only want to generate the CFG view，how to use it

**Comment by YoshikiHigo** [comment]
Thanks to everyone who asked here, and sorry for the long silence. Two things have changed since this thread.

- The README now describes each command line tool with its options: https://github.com/YoshikiHigo/TinyPDG#command-line-tools. To get only the CFG, run Writer with `-c`: `bin/writer -d <source directory> -c cfg.dot` (add `-p pdg.dot` if you also want the PDG). Render the file with Graphviz, for example `dot -Tpdf cfg.dot -o cfg.pdf`.
- The jar produced by Gradle was not runnable on its own, as noted above. `./gradlew installDist` (#14) now creates `build/install/TinyPDG/bin/` with a start script per tool (`writer`, `scorpio`, `distiller`, `predictor`, plus `.bat` files for Windows) and all the jars they need in `lib/`, so no classpath has to be assembled by hand. `./gradlew build` also produces the same layout as a zip under `build/distributions/`.

Closing this issue. If something in the README is unclear, please open a new one saying which step failed.
